Panic in pkg/operator/vttablet/pvc.go

A user has reported a panic with this stack trace:

```
1 runtime.go:69] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190221213512-86fb29eff628/pkg/util/runtime/runtime.go:76
/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190221213512-86fb29eff628/pkg/util/runtime/runtime.go:65
/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190221213512-86fb29eff628/pkg/util/runtime/runtime.go:51
/usr/local/go/src/runtime/panic.go:679
/usr/local/go/src/runtime/panic.go:199
/usr/local/go/src/runtime/signal_unix.go:394
/go/src/planetscale.dev/vitess-operator/pkg/operator/vttablet/pvc.go:34
/go/src/planetscale.dev/vitess-operator/pkg/controller/vitessshard/reconcile_backup_job.go:156
/go/src/planetscale.dev/vitess-operator/pkg/operator/reconciler/object.go:166
/go/src/planetscale.dev/vitess-operator/pkg/operator/reconciler/object_set.go:87
/go/src/planetscale.dev/vitess-operator/pkg/controller/vitessshard/reconcile_backup_job.go:152
/go/src/planetscale.dev/vitess-operator/pkg/controller/vitessshard/vitessshard_controller.go:198
/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.1.12/pkg/internal/controller/controller.go:215
/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.1.12/pkg/internal/controller/controller.go:158
/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190221213512-86fb29eff628/pkg/util/wait/wait.go:133
/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190221213512-86fb29eff628/pkg/util/wait/wait.go:134
/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190221213512-86fb29eff628/pkg/util/wait/wait.go:88
/usr/local/go/src/runtime/asm_amd64.s:1357
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1b0bf43]
```

--- a/pkg/controller/vitessshard/reconcile_backup_job.go
+++ b/pkg/controller/vitessshard/reconcile_backup_job.go
@@ -76,9 +76,10 @@ func (r *ReconcileVitessShard) reconcileBackupJob(ctx context.Context, vts *plan
 	// Here we only care about complete backups.
 	completeBackups := vitessbackup.CompleteBackups(allBackups.Items)
 
-	// Generate keys (object names) for all desired backup jobs.
+	// Generate keys (object names) for all desired backup Pods and PVCs.
 	// Keep a map back from generated names to the backup specs.
-	keys := []client.ObjectKey{}
+	podKeys := []client.ObjectKey{}
+	pvcKeys := []client.ObjectKey{}
 	specMap := map[client.ObjectKey]*vttablet.BackupSpec{}
 
 	// The object name for the initial backup Pod, if we end up needing one.
@@ -96,7 +97,10 @@ func (r *ReconcileVitessShard) reconcileBackupJob(ctx context.Context, vts *plan
 		// nothing and return success.
 		initSpec := vtbackupInitSpec(initPodKey, vts, labels)
 		if initSpec != nil {
-			keys = append(keys, initPodKey)
+			podKeys = append(podKeys, initPodKey)
+			if initSpec.TabletSpec.DataVolumePVCSpec != nil {
+				pvcKeys = append(pvcKeys, initPodKey)
+			}
 			specMap[initPodKey] = initSpec
 		}
 	} else {
@@ -106,7 +110,7 @@ func (r *ReconcileVitessShard) reconcileBackupJob(ctx context.Context, vts *plan
 
 	// Reconcile vtbackup Pods.
 	orphanPods := map[client.ObjectKey]*corev1.Pod{}
-	err := r.reconciler.ReconcileObjectSet(ctx, vts, keys, labels, reconciler.Strategy{
+	err := r.reconciler.ReconcileObjectSet(ctx, vts, podKeys, labels, reconciler.Strategy{
 		Kind: &corev1.Pod{},
 
 		New: func(key client.ObjectKey) runtime.Object {
@@ -148,8 +152,9 @@ func (r *ReconcileVitessShard) reconcileBackupJob(ctx context.Context, vts *plan
 		resultBuilder.Error(err)
 	}
 
-	// Reconcile vtbackup PVCs. Use the same key as the corresponding Pod.
-	err = r.reconciler.ReconcileObjectSet(ctx, vts, keys, labels, reconciler.Strategy{
+	// Reconcile vtbackup PVCs. Use the same key as the corresponding Pod,
+	// but only if the Pod expects a PVC.
+	err = r.reconciler.ReconcileObjectSet(ctx, vts, pvcKeys, labels, reconciler.Strategy{
 		Kind: &corev1.PersistentVolumeClaim{},
 
 		New: func(key client.ObjectKey) runtime.Object {

